backend/src/main/scala/bloop/task/Task.scala

```diff
@@ -0,0 +1,668 @@
+            val t = task.runAsync(sh)
+            t.onComplete {
+              case Failure(e) => completeWithError(e)
+              case Success(a) => result(a, idx)
+            }(sh)
+            t
+        }
+        Cancelable { () =>
+          all.foreach(_.cancel())
+        }
+      }
+    }
+  }
+
+  def gatherUnordered[A](in: Iterable[Task[A]]): Task[List[A]] =
+    parSequenceUnordered(in)
+
+  def parSequenceUnordered[A](in: Iterable[Task[A]]): Task[List[A]] = {
+    val size = in.size
+    if (in.isEmpty) {
+      Task.now(List.empty)
+    } else {
+      Task.create[List[A]] { (sh, cb) =>
+        val state = new AtomicReference(Vector.empty[A])
+
+        def completeWithError(e: Throwable): Unit = {
+          state.get() match {
+            case null =>
+            case curr =>
+              if (state.compareAndSet(curr, null))
+                cb.onError(e)
+              else
+                completeWithError(e)
+          }
+        }
+
+        def result(a: A): Unit = {
+          state.get() match {
+            case null =>
+            case acc =>
+              val next = acc :+ a
+              if (next.length == size)
+                cb.onSuccess(next.toList)
+              else if (!state.compareAndSet(acc, next))
+                result(a)
+          }
+        }
+
+        val all = in.map { task =>
+          val t = task.runAsync(sh)
+          t.onComplete {
+            case Failure(e) => completeWithError(e)
+            case Success(a) => result(a)
+          }(sh)
+          t
+        }
+        Cancelable { () =>
+          all.foreach(_.cancel())
+        }
+      }
+    }
+  }
+
+  def parSequenceN[A](n: Int)(in: Iterable[Task[A]]): Task[List[A]] = {
+    val chunks = in.grouped(n).toList.map(group => Task.parSequence(group))
+    Task.sequence(chunks).map(_.flatten)
+  }
+
+  def fromFuture[A](f: Future[A]): Task[A] =
+    Wrap(MonixTask.fromFuture(f), List.empty)
+
+  def zip2[A, B](a: Task[A], b: Task[B]): Task[(A, B)] =
+    mapBoth(a, b) { case (a, b) => (a, b) }
+
+  def mapBoth[A, B, R](a: Task[A], b: Task[B])(f: (A, B) => R): Task[R] = {
+    Task
+      .create[(A, B)] { (sh, cb) =>
+        val result = new AtomicReference[(Option[A], Option[B])]((None, None))
+        val fa = a.runAsync(sh)
+        val fb = b.runAsync(sh)
+
+        def aResult(a: A): Unit = {
+          result.get() match {
+            case (None, Some(b)) =>
+              result.set(null)
+              cb.onSuccess((a, b))
+            case curr @ (None, None) =>
+              if (!result.compareAndSet(curr, (Some(a), None))) {
+                aResult(a)
+              }
+            case _ =>
+          }
+        }
+        def bResult(b: B): Unit = {
+          result.get() match {
+            case (Some(a), None) =>
+              result.set(null)
+              cb.onSuccess((a, b))
+            case curr @ (None, None) =>
+              if (!result.compareAndSet(curr, (None, Some(b)))) {
```

Comment:
> Isn't this a raceCondition if both end at the same time?

no - they can end at the same time but at the single moment of time only one can update the result by `compareAndSet`.
So, in this case - one performs `compareAndSet` while second doesn't because `curr` becomes outdated and then it executes this function one more time and completes task.